python/ray/util/client/common.py

```diff
@@ -402,3 +418,241 @@ def _get_client_id_from_context(context: Any) -> str:
         logger.error("Client connecting with no client_id")
         context.set_code(grpc.StatusCode.FAILED_PRECONDITION)
     return client_id
+
+
+def _propagate_error_in_context(e: Exception, context: Any) -> bool:
+    """
+    Encode an error into the context of an RPC response. Returns True
+    if the error can be recovered from, false otherwise
+    """
+    try:
+        if isinstance(e, grpc.RpcError):
+            # RPC error, propagate directly by copying details into context
+            context.set_code(e.code())
+            context.set_details(e.details())
+            return e.code() not in GRPC_UNRECOVERABLE_ERRORS
+    except Exception:
+        # Extra precaution -- if encoding the RPC directly fails fallback
+        # to treating it as a regular error
+        pass
+    context.set_code(grpc.StatusCode.FAILED_PRECONDITION)
+    context.set_details(str(e))
+    return False
+
+
+def _id_is_newer(id1: int, id2: int) -> bool:
```

Comment:
confused, why do we need this? Is this because of int overflow? Firstly, if it's int64, usually it's ok and we don't need this trick.
And besides, I think for python int is unbounded? (https://stackoverflow.com/questions/7604966/maximum-and-minimum-values-for-ints)

Reply:
The IDs eventually get serialized to protobufs for gRPC, which are bounded to int32 at the moment

Reply:
Should we use uint64 instead of int32 then?